### PR TITLE
Refine `Either.parseJSON` return type

### PR DESCRIPTION
**Pull request by OliverJAsh**
Fixes https://github.com/gcanti/fp-ts/issues/1251

### CODE REVIEW

File: src/Either.ts [hunk 1]
```diff
@@ -169,6 +169,13 @@ export function tryCatch<E, A>(f: Lazy<A>, onError: (e: unknown) => E): Either<E
   }
 }
 
+// Copied from https://github.com/Microsoft/TypeScript/issues/1897#issuecomment-338650717
+type AnyJson = boolean | number | string | null | JsonArray | JsonMap
+interface JsonMap {
+  readonly [key: string]: AnyJson
```
Comment (OliverJAsh) [inline src/Either.ts:175]: I made this `readonly` because the lint rule told me to. 🤷‍♂️

File: src/Either.ts [hunk 2]
```diff
@@ -169,6 +169,13 @@ export function tryCatch<E, A>(f: Lazy<A>, onError: (e: unknown) => E): Either<E
   }
 }
 
+// Copied from https://github.com/Microsoft/TypeScript/issues/1897#issuecomment-338650717
+type AnyJson = boolean | number | string | null | JsonArray | JsonMap
+interface JsonMap {
```
Comment (OliverJAsh) [inline src/Either.ts:174]: I initially use `type` but recursive type aliases are only supported in newer TS versions, so had to revert to `interface`.

**Comment by OliverJAsh** [comment]
Sure, let me do that.

> * rename `JsonMap` to `JsonRecord` for consistency with the built-in `Record` type

Note: I still feel uncomfortable using "record" here (https://github.com/gcanti/fp-ts/issues/685).

**Comment by OliverJAsh** [comment]
@gcanti Maybe we should export them too, WDYT?

**Comment by gcanti** [comment]
@OliverJAsh yes, that's a useful type which [I would like to import somewhere else](https://github.com/gcanti/io-ts/blob/c85564afc304b852033e0dc1b99ca22cea920afa/src/JsonEncoder.ts#L13-L31)

File: docs/modules/Either.ts.md [hunk 3]
```diff
@@ -417,7 +417,7 @@ Converts a JavaScript Object Notation (JSON) string into an object.
 **Signature**
 
 ```ts
-export declare function parseJSON<E>(s: string, onError: (reason: unknown) => E): Either<E, unknown>
+export declare function parseJSON<E>(s: string, onError: (reason: unknown) => E): Either<E, AnyJson>
```
Comment (gcanti) [inline docs/modules/Either.ts.md:420]: looks like the generated docs are stale

**Comment by awerlogus** [comment]
I think, you should add also `| readonly []` in `AnyJson` because https://github.com/microsoft/TypeScript/issues/38727#issuecomment-643495070

**Comment by gcanti** [comment]
@awerlogus why? all tuples, `readonly []` included, are already modeled by `JsonArray`

**Comment by awerlogus** [comment]
@gcanti compare the types of variables `a` and `b`.

```ts
type Json1 = boolean | number | string | null | JsonArray1 | JsonRecord1

interface JsonArray1 extends ReadonlyArray<Json1> { }

export interface JsonRecord1 {
  readonly [key: string]: Json1
}

function handleJson1<T extends Json1>(arg: T): T {
  return arg
}


type Json2 = boolean | number | string | null | JsonArray2 | JsonRecord2 | readonly []

interface JsonArray2 extends ReadonlyArray<Json2> { }

export interface JsonRecord2 {
  readonly [key: string]: Json2
}

function handleJson2<T extends Json2>(arg: T): T {
  return arg
}

// number[]
const a = handleJson1([1, 2, 3])

// [number, number, number]
const b = handleJson2([1, 2, 3])
```

Type parameters will be not inferred as tuples if the type that extends them doesn't contain `readonly []` (or another tuple) explicitly. That's why I always combine `ReadonlyArray` with `readonly []`.

**Comment by gcanti** [comment]
@awerlogus ok but `handleJson` is just the identity function, could you show a compelling use case where `Json2` will be actually useful?

**Comment by awerlogus** [comment]
@gcanti 

> could you show a compelling use case where `Json2` will be actually useful?

You're requesting for a json processing example? It obviously will be absolutely synthetic. Somebody may be needed to handle tuples and simple arrays differently on type level. And it will be better if types will be inferred as narrow as possible. That's all. Adding `readonly []` is not such a serious suggestion, which will determine the development of humanity for years to come, to be taken with such hostility.

**Comment by gcanti** [comment]
@awerlogus there's no hostility in my request, I'm genuinely interested in a use case since I never thought of adding `readonly []` to the untagged union (as linked in my comment [here](https://github.com/gcanti/fp-ts/pull/1252#issuecomment-649204318)) and I may be missing something important.

**Comment by awerlogus** [comment]
@gcanti as I said, the example turned out to be synthetic. Here you can see that type inference in case of `getJsonLength2` provides us more info to operate with on type level.

```ts
type Json1 = boolean | number | string | null | JsonArray1 | JsonRecord1

interface JsonArray1 extends ReadonlyArray<Json1> { }

export interface JsonRecord1 {
  readonly [key: string]: Json1
}

declare function getJsonLength1<T extends Json1>(arg: T): T extends ReadonlyArray<any> ? T['length'] : 1


type Json2 = boolean | number | string | null | JsonArray2 | JsonRecord2 | readonly []

interface JsonArray2 extends ReadonlyArray<Json2> { }

export interface JsonRecord2 {
  readonly [key: string]: Json2
}

declare function getJsonLength2<T extends Json2>(arg: T): T extends ReadonlyArray<any> ? T['length'] : 1


// number
const a = getJsonLength1([1, 2, 3])

// 3
const b = getJsonLength2([1, 2, 3])


// 1
const c = getJsonLength1({ a: 'foo', b: 'bar' })

// 1
const d = getJsonLength2({ a: 'foo', b: 'bar' })
```

**Pull request merged**